components/tidb_query/src/codec/mysql/json/binary.rs

```diff
@@ -218,170 +293,30 @@ impl<T: BufferWriter> JsonEncoder for T {}
-        let buf = self.bytes();
-        if buf.len() < left_size {
-            return Err(Error::unexpected_eof());
-        }
-        let buf = &buf[..left_size];
-        let value_entries_len = VALUE_ENTRY_LEN * element_count;
-        let (mut value_entries_data, values_data) = buf.split_at(value_entries_len);
-        let mut array_data = Vec::with_capacity(element_count);
-        let data_start_offset = (U32_LEN + U32_LEN + value_entries_len) as u32;
-        for _ in 0..element_count {
-            let value = value_entries_data.read_json_item(values_data, data_start_offset)?;
-            array_data.push(value);
-        }
-        self.advance(left_size);
-        Ok(Json::Array(array_data))
-    }
-
-    fn read_json_str(&mut self) -> Result<Json> {
-        let length = self.read_var_u64()? as usize;
-        let encode_value = self.read_bytes(length)?;
-        let value = str::from_utf8(encode_value)?.to_owned();
-        Ok(Json::String(value))
-    }
-
-    fn read_json_literal(&mut self) -> Result<Json> {
-        match self.read_u8()? {
-            JSON_LITERAL_TRUE => Ok(Json::Boolean(true)),
-            JSON_LITERAL_FALSE => Ok(Json::Boolean(false)),
-            _ => Ok(Json::None),
-        }
-    }
-
-    fn read_json_double(&mut self) -> Result<Json> {
-        let value = self.read_f64_le()?;
-        Ok(Json::Double(value))
-    }
-
-    fn read_json_i64(&mut self) -> Result<Json> {
-        let value = self.read_i64_le()?;
-        Ok(Json::I64(value))
-    }
-
-    fn read_json_u64(&mut self) -> Result<Json> {
-        let value = self.read_u64_le()?;
-        Ok(Json::U64(value))
-    }
-
-    fn read_json_item(&mut self, values_data: &[u8], data_start_position: u32) -> Result<Json> {
-        let mut entry = self.read_bytes(VALUE_ENTRY_LEN)?;
-        let code = entry.read_u8()?;
-        match code {
-            TYPE_CODE_LITERAL => entry.read_json_literal(),
-            _ => {
-                let real_offset = entry.read_u32_le()?;
-                let offset_in_values = real_offset - data_start_position;
-                let mut value = &values_data[offset_in_values as usize..];
-                value.read_json_body(code)
+        let tp: JsonType = self.read_u8()?.into();
+        let value = match tp {
+            JsonType::Object | JsonType::Array => {
+                let value = self.bytes();
+                let data_size = NumberCodec::decode_u32_le(&value[ELEMENT_COUNT_LEN..]) as usize;
+                self.read_bytes(data_size)?
             }
-        }
+            JsonType::String => {
+                let value = self.bytes();
+                let (str_len, len_len) = NumberCodec::try_decode_var_u64(&value).unwrap();
+                self.read_bytes(str_len as usize + len_len)?
+            }
+            JsonType::I64 | JsonType::U64 | JsonType::Double => self.read_bytes(NUMBER_LEN)?,
+            JsonType::Literal => self.read_bytes(LITERAL_LEN)?,
```

Comment:
Four bytes has been written in L293 (https://github.com/tikv/tikv/pull/6388/files#diff-65ef8b5159dd7aec0e01ebe808e73f9dR273), but only read one byte, is it correct?

Reply:
Here the reading is for a plain `Literal` value (not in an Object or Array) and it's not padded.